### ISSUE TITLE
Fix or ignore new core lints in generated code

### ISSUE THREAD

**Issue by jakemac53** [google/built_value.dart#1115]
After updating to package:lints version 2.0.0, we get a few new lints in generated built_value code.

The specific lints are:

- library_private_types_in_public_api
- no_leading_underscores_for_local_identifiers

**Comment by natebosch** [comment]
What were the specifics of the `library_private_types_in_public_api`? Is this something we should fix in `built_value`? Or fix in the linter?

**Comment by jakemac53** [comment]
Example here https://github.com/dart-lang/build/blob/06be5da4a847d390755a7aa1eca2e010879de496/build_daemon/lib/data/build_request.g.dart#L80

This might be able to be the less specific interface that the generated private type is implementing (`BuildRequest`). Or the lint could just be ignored.

**Comment by natebosch** [comment]
I think this is worth fixing.
This does show up unnecessarily in the docs.
https://pub.dev/documentation/build_daemon/latest/data_build_request/BuildRequestBuilder/build.html

It would mean adding a cast in the factory constructor. https://github.com/dart-lang/build/blob/06be5da4a847d390755a7aa1eca2e010879de496/build_daemon/lib/data/build_request.g.dart#L34

We already rely on casts.
https://github.com/dart-lang/build/blob/06be5da4a847d390755a7aa1eca2e010879de496/build_daemon/lib/data/build_request.g.dart#L71

**Comment by jakemac53** [comment]
> It would mean adding a cast in the factory constructor. https://github.com/dart-lang/build/blob/06be5da4a847d390755a7aa1eca2e010879de496/build_daemon/lib/data/build_request.g.dart#L34

Or just change this to a static method? That would be breaking though (couldn't use a delegating constructor).

**Comment by natebosch** [comment]
> That would be breaking though (couldn't use a delegating constructor)

Yeah, that's how it's intended to be used today so changing it would require a migration for all use of `built_value`

**Comment by jakemac53** [comment]
Ultimately with macros we can just generate the factory constructor and be happy, for now a cast sounds fine.

**Comment by davidmorgan** [comment]
Thanks Nate, thanks Jake.

This is a weird corner. The generated "build" method is saying "I return the generated impl of the class", which it does, and should; you're not allowed to subclass to provide something else. But we want the generated impl class to be private because you should always refer to the public interface, and so the checked in code can control instantiation.

I'm not convinced the interface change is an improvement--it's less confusing, sure, but also less accurate. And the cast might introduce a performance overhead.

tl;dr: I suggest adding an ignore, this is a problem unique to generation that goes away with macros.

no_leading_underscores_for_local_identifiers also seems like it should get an ignore, the leading underscore is to avoid a clash between public fields and generated locals, which seems a reasonable thing for generators to do.

WDYT please?

Thanks.

**Comment by natebosch** [comment]
Yeah I think today we're just counting on the private class not having any new public members (It shouldn't need to anyway).

Another option is to introduce yet another private class to hide it. I'm not sure if there are performance implications.

```dart
abstract class BuildRequestBuilder
    implements Builder<BuildRequest, BuildRequestBuilder> {
    BuildRequestBuilder() = _BuildRequestBuilder
}

class _BuildRequestBuilder implements BuildRequestBuilder {
    @override
    _$BuildRequest build() {
```

The only place that needs to see the `_$BuildRequest` return type could still use a `_BuildRequestBuilder` instead of the public interface because it is instantiating directly.

**Comment by davidmorgan** [comment]
Hmm we could also do

```
BuildRequest build() => _build();

_$BuildRequest _build() {
...
}
```

and call the private one from the factory constructor, I'll give that a try.

**Comment by davidmorgan** [comment]
https://github.com/google/built_value.dart/pull/1139

WDYT, please? @jakemac53 @natebosch
